### PR TITLE
Add align_offset intrinsic

### PR DESCRIPTION
**Pull request by oli-obk**
see https://github.com/rust-lang/rfcs/pull/2043 for details and the plan towards stabilization (reexport in `core::mem` via various convenience functions)

as per @scottmcm 's [comment](https://github.com/rust-lang/rfcs/pull/2043#issuecomment-316818169), this is just the intrinsic (which is obviously unstable).

### CODE REVIEW

File: src/libcore/intrinsics.rs [hunk 1]
```diff
@@ -1343,4 +1343,35 @@ extern "rust-intrinsic" {
...
+/// Computes the byte offset that needs to be applied to `ptr` in order to
+/// make it aligned to `align`.
+/// If it is not possible to align `ptr`, the implementation returns
+/// `usize::max_value()`.
+///
+/// There are no guarantees whatsover that offsetting the pointer will not
+/// overflow or go beyond the allocation that `ptr` points into.
+/// It is up to the caller to ensure that the returned offset is correct
+/// in all terms other than alignment.
+pub fn align_offset(ptr: *const (), align: usize) -> usize {
```
Comment (scottmcm) [inline src/libcore/intrinsics.rs]: nit: Should this be unsafe, for consistency with the real intrinsic?

File: src/libcore/intrinsics.rs [hunk 2]
```diff
@@ -1343,4 +1343,35 @@ extern "rust-intrinsic" {
...
+
+    /// Computes the byte offset that needs to be applied to `ptr` in order to
+    /// make it aligned to `align`.
+    /// If it is not possible to align `ptr`, the implementation returns
+    /// `usize::max_value()`.
+    ///
+    /// There are no guarantees whatsover that offsetting the pointer will not
+    /// overflow or go beyond the allocation that `ptr` points into.
+    /// It is up to the caller to ensure that the returned offset is correct
+    /// in all terms other than alignment.
```
Comment (scottmcm) [inline src/libcore/intrinsics.rs]: Consider adding an `# Examples` doctest to demo a few simple cases.

File: src/libcore/intrinsics.rs [hunk 3]
```diff
@@ -1343,4 +1343,75 @@ extern "rust-intrinsic" {
...
+    ///
+    /// Accessing adjacent `u8` as `u16`
+    ///
+    /// ```
+    /// # let n = 3;
+    /// # use std::intrinsics::align_offset;
+    /// # use std::mem::align_of;
+    /// let x = [5u8, 6u8, 7u8, 8u8, 9u8];
+    /// let ptr = &x[n] as *const u8;
+    /// let offset = align_offset(ptr as *const (), align_of::<u16>());
```
Comment (kennytm) [inline src/libcore/intrinsics.rs]: doc test failed.

```rust
[00:56:48] failures:
[00:56:48] 
[00:56:48] ---- intrinsics.rs - intrinsics::align_offset (line 1361) stdout ----
[00:56:48] 	error: use of unstable library feature 'core_intrinsics': intrinsics are unlikely to ever be stabilized, instead they should be used through stabilized interfaces in the rest of the standard library (see issue #0)
[00:56:48]  --> intrinsics.rs:9:14
[00:56:48]   |
[00:56:48] 9 | let offset = align_offset(ptr as *const (), align_of::<u16>());
[00:56:48]   |              ^^^^^^^^^^^^
[00:56:48]   |
[00:56:48]   = help: add #![feature(core_intrinsics)] to the crate attributes to enable
[00:56:48] 
[00:56:48] error: use of unstable library feature 'core_intrinsics': intrinsics are unlikely to ever be stabilized, instead they should be used through stabilized interfaces in the rest of the standard library (see issue #0)
[00:56:48]  --> intrinsics.rs:5:5
[00:56:48]   |
[00:56:48] 5 | use std::intrinsics::align_offset;
[00:56:48]   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[00:56:48]   |
[00:56:48]   = help: add #![feature(core_intrinsics)] to the crate attributes to enable
[00:56:48] 
[00:56:48] error[E0308]: mismatched types
[00:56:48]   --> intrinsics.rs:11:30
[00:56:48]    |
[00:56:48] 11 |     let u16_ptr = ptr.offset(offset) as *const u16;
[00:56:48]    |                              ^^^^^^ expected isize, found usize
[00:56:48] 
[00:56:48] thread 'rustc' panicked at 'couldn't compile the test', /checkout/src/librustdoc/test.rs:280:12
[00:56:48] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:56:48] 
[00:56:48] 
[00:56:48] failures:
[00:56:48]     intrinsics.rs - intrinsics::align_offset (line 1361)
[00:56:48] 
[00:56:48] test result: FAILED. 1231 passed; 1 failed; 1 ignored; 0 measured; 0 filtered out
[00:56:48] 
[00:56:48] error: test failed, to rerun pass '--doc'
```

**Comment by oli-obk** [comment]
ping @aturon

**Pull request merged**